Cloud Init 'nocloud' datasource support

davidvossel opened the pull request:
**Read the doc provided in docs/cloud-init.md for examples and more information about cloud-init**

This is a redesign of #340.  These patches establish a pattern for injecting disk based cloud init datasources into vms. The 'nocloud' datasource is implemented to both prove the pattern works and allow us to begin iterating on the node isolator use case using fedora-atomic images (which support the nocloud datasource). Future disk based data sources (like configDrive) should be easy to add later on.

This is the high level flow of events for how nocloud data is injected into a VM.
- Cloud init user-data is placed on the vm definition and posted to the cluster.
- virt-handler posts the CloudInitSpec to the new virt-config-disk component to generate the iso right before the VM launches.
- The disks generated by virt-config-disk are shared with the libvirt process using a shared file mount. Each domain gets a cloud init subdirectory that is shared between libvirt and virt-config-disk.
- Virt-handler injects a reference to the cloud-init disk in the domain xml.
- The VM launches with access to the cloud-init data provided by a block device.

michalskrivanek commented:
> This is the high level flow of events for how nocloud data is injected into a VM.
> 
>     Cloud init user-data is placed on the vm definition and posted to the cluster.

in #340 rmohr sugested a separate TPR. 

>     virt-controller passes the userdata on the VM spec and other cloud init information to the virt-launcher pod via enviroment variables

why virt-launcher and not virt-handler?

>     virt-launcher generates the iso disk file in a way that adheres to the cloud-init data source standard in use.

So in general there are two options, either do it the "quick and dirty" way and create the ISO at the time of launch like you propose, or delegate it into independent process handled by virt-controller and perhaps a dedicated service. 
The disadvantage of virt-launcher is that you're really at the lowest layer and you build a complex and potentially a bit large thing during the time-sensitive VM launch. Advantage is that you do not need to coordinate the ISO creation and VM/POD lifecycle separately.

>     The disks generated by virt-launcher are shared with the libvirt process using a shared file mount. Each domain gets a cloud init subdirectory that is shared between libvirt and virt-launcher.
>     Virt-handler injects a reference to the cloud-init disk in the domain xml.

I suppose you mean the backend part. On the guest-facing side it needs to fit into other disks

>     The VM launches with access to the cloud-init data provided by a block device.

Hm, if you plan to supply the disks via mount, why bother with building the ISO? Can't we just create the file(s)
And as commented in #340 I still see it as a problem that it's called cloudinit everywhere. There are equally important things we need to supply via the same mechanism, like windows sysprep and windows drivers

davidvossel commented:
> in #340 rmohr sugested a separate TPR.

I think k8s secrets make more sense. I consider this an additional feature though built on what we have here. people would have the option to include user data in the VM definition, or reference userdata placed in a k8s secret in the VM definition

Example: Create a k8s secret called my-user-data-secret and reference that secret in the cloud init VM definition.
```
metadata:
  name: atomic-ephemeral
apiVersion: kubevirt.io/v1alpha1
kind: VM
spec:
  domain:
    devices:
      disks:
      - type: ContainerRegistryDisk:v1alpha
        source:
          name: kubevirt/fedora-atomic-registry-disk-demo:devel
        target:
          dev: vda
      interfaces:
      - source:
          network: default
        type: network
    memory:
      unit: MB
      value: 1024
    os:
      type:
        os: hvm
    type: qemu
  cloudInit:
    dataSource: noCloud
    noCloudData:
      userDataSecretRef: my-user-data-secret
      diskTarget: vdb

```

> So in general there are two options, either do it the "quick and dirty" way and create the ISO at the time of launch like you propose, or delegate it into independent process handled by virt-controller and perhaps a dedicated service.

This actually is handled by virt-controller in a deliberate manner that avoids race conditions.  I wouldn't necessarily call this "quick and dirty". It is done in a controlled way by a dedicated process before the VM is launched.  

virt-controller now adds a readiness check to the virt-launcher container. This ready check ensures that virt-controller will not schedule the VM onto a node until virt-launcher has completed initializing. So... virt-controller now waits on virt-launcher to initalize before allowing the VM to be placed on the node.

admiyo commented:
retest this please

davidvossel commented:
> Hm, if you plan to supply the disks via mount, why bother with building the ISO? Can't we just create the file(s)

We're adhering to a cloud-init spec called 'nocloud'. That specification requires a block device with a specific volume label. To meet this specification, we don't have the liberty to inject these files however we want. The cloud-init processes inside the VM need them formatted in a specific way.

> And as commented in #340 I still see it as a problem that it's called cloudinit everywhere. There are equally important things we need to supply via the same mechanism, like windows sysprep and windows drivers

It's called cloud-init everywhere because that's the name of the project we're supporting. https://cloudinit.readthedocs.io/en/latest/ Cloud-init is a set of processes that live in the VM that can interpret userdata and metadata injected into a VM via multiple clould-init data sources. https://cloudinit.readthedocs.io/en/latest/topics/datasources.html#datasource-documentation

The functionality you're referring to isn't related to any cloud-init data source specification. Injecting something like a windows driver binary into a VM is separate from the dynamically genereated cloud-init userdata.

I'd expect users to provide drivers to a VM by explicitly attaching a disk to the VM with those files on them.  The custom cloud-init userdata on a VM would be used to install/provision the VM.  That userdata script could actually be responsible for finding the binaries on another disk and installing them locally on the VM.

davidvossel commented:
Here's an update.

After some discussion with @fabiand and @stu-gott we decided it might be a good idea to shift the nocloud iso generation into a side car process in the libvirt pod rather than having virt-launcher responsible for generating this data.

The concern is trust.  We don't want to trust virt-launcher to directly generate and share data that is consumed by the VM.

davidvossel commented:
@fabiand @michalskrivanek @stu-gott 

### Update

I broke out the iso creation into a separate "generic" component called **virt-config-disk**

### New virt-config-disk component

This new component is responsible for generating local disks containing configuration data for a vm. This could be a disk for a cloud-init provider, or something else entirely. 

The component is completely self contained and makes no connection out to the rest of the cluster.  'Specs' are posted to the component and in turn the component converts the spec to a disk.  Libvirt and virt-config-disk have a shared host directory that makes sharing these ISOs with the qemu processes possible.

For example, a CloudInitSpec object can be posted to the component and it will generate the corresponding iso.  In the future this component can be used to generate local data for other specs. Although it only has an implementation for cloud-init specs right now, it is not intended to be limited in scope to only cloud-init. 

### API

The virt-config-disk component opens a unix domain socket that it accepts HTTP traffic over.  The virt-handler component has permission to access this socket.

The API is simple and can be extended easily. 
```
POST /<action>/<data type>/<namespace>/<domain>
```
The body is expected to be a json representation of the 'data type'

Example: post cloudinit data to the service for my-vm in my-namespace
POST /create/cloudinit/my-namespace/my-vm with the CloudInitSpec field encoded a json in the body.

### Where this fits in the VM launch flow

Unlike our original design, the virt-config-disk component has no relationship with virt-launcher. Virt-handler invokes virt-config-disk before the VM is allowed to start and after the VM is killed.  The VM start action is still dependent on virt-config-disk passing and the local data being generated first.

The reason for this change is I did not want virt-config-disk to have permission to talk to the k8s cluster. I wanted it be completely self contained and only react to the data that was provided to it.  If virt-launcher invoked virt-config-disk to generate the local data, then virt-config-disk would have had to retrieve the specs from k8s.  Virt-handler already has the spec information in cache, and can pass it directly to virt-config-disk without requiring additional k8s api calls. 

### Remaining work

I plan to squash commits and introduce unit testing for the virt-config-disk component shortly. I just wanted to post up what I have now for an initial round of feedback.

davidvossel commented:
Alrighty, here's another update.

- rebased to trunk to fix merge conflicts
- reorganized and squashed the commits so they make more sense
- added virt-config-disk unit tests to the config-disk pkg that exercise the client/server interaction and do mock iso generation.

davidvossel commented:
retest this please

davidvossel commented:
Made another update. Added the following
- cloud-init targeted unit tests.
- made metadata auto-generation optional
- improved error detection. invalid folder permissions will result in a fast fail now. 
- explicitly define host share directories in virt-handler manifest rather than depending on defaults. I thing this makes it more obvious where things are located.

fabiand commented on cluster/vm-nocloud.yaml:
```diff
@@ -0,0 +1,29 @@
+      type:
+        os: hvm
+    type: qemu
+  cloudInit:
```
Looking at this example, I'd really go with the disk approach.

If we go with the disk approach, then we don't clobber the spec, and the user got explicit conroll (and this is what VM API is about, explicit controll) over where the disk is located.
VMConfig could be the place where cloud-init is exposed as a`feature`.

Thus:
```
disk:
- type: CloudInit
  mode: noCloud # needed?
  userData: …
  target:
    dev: vda
```

davidvossel replied:
> Looking at this example, I'd really go with the disk approach.

I disagree. The 'nocloud' datasource happens to use a disk. There are other cloud-init data sources that will not.  

I'd rather not see this configuration data as something that is intermingled with the VM spec. These are logically two separate concepts.  The VM spec defines what the domain looks like, The cloud-init spec defines the cloud-init datasource method used to inject configuration into the the VM.  The latter could involve injecting a disk into the VM, or it could involve talking to a meta-data server. 

> f we go with the disk approach, then we don't clobber the spec, and the user got explicit conroll (and this is what VM API is about, explicit controll) over where the disk is located.
VMConfig could be the place where cloud-init is exposed as a feature.

I'm doubtful that we'll ever care about where a disk is placed on a vm that uses cloud-init and the nocloud data source.

Can you give an example of when disk placement matters?

michalskrivanek replied:
the disk naming needs to be predictable to the guest with or without cloud-init. The only way to make sure it fits within other disks is to make it part of the spec. The actual content can be separate, but there should be a disk device referring to it in the spec section

davidvossel replied:
> the disk naming needs to be predictable to the guest with or without cloud-init. The only way to make sure it fits within other disks is to make it part of the spec. The actual content can be separate, but there should be a disk device referring to it in the spec section

This coordination already happens, but it is the inverse of what you're proposing. The CloudInit spec's 'nocloud' structure takes in the 'diskTarget' name that is used when injecting the disk into the domain xml.

fabiand replied:
> I disagree. The 'nocloud' datasource happens to use a disk. There are other cloud-init data sources that will not.

But there is only one noCloud source, and this is disk based iiuic.



> I'd rather not see this configuration data as something that is intermingled with the VM spec. These are logically two separate concepts.


> I'm doubtful that we'll ever care about where a disk is placed on a vm that uses cloud-init and the nocloud data source.

We do need to care about this, for some operating system (Windows) it is important what disk it is, and what it's label is - so we do need to have control.

By making it a disk, a consumer (like us) has the explciti control over where the disk is placed.



> This coordination already happens, but it is the inverse of what you're proposing. The CloudInit spec's 'nocloud' structure takes in the 'diskTarget' name that is used when injecting the disk into the domain xml.

The disk target field is not recognized by all OSes we can not depend on it. Bus/device addresses are relevant. These could be implicitly defined by the ordering of the devices - but the disk target is not reliable.

davidvossel replied:
> The disk target field is not recognized by all OSes we can not depend on it. Bus/device addresses are relevant. These could be implicitly defined by the ordering of the devices - but the disk target is not reliable.

@fabiand I'm fairly certain the diskTarget is going to be recognized by all OSes that care about the nocloud data source.... or cloud-init at all for that matter.  

I know the Windows use case where we have to care about disk ordering and need explicit control. That use case isn't applicable here. cloud-init is not a windows thing.

michalskrivanek replied:
I believe there is an integration with Openstack for https://cloudbase.it/cloudbase-init/ ?

michalskrivanek replied:
@davidvossel it's not only about recognizing the nocloud disk (I trust you it works today, though I still have a vivid memory of really crappy cloud-init versions in RHEL 6) but equally about not screwing up other disks. You do not want the other disks to "shift".  If the disk is explicitly in domain spec you it's more clear how it affects the rest of the VM

rmohr replied:
We were discussing this yesterday offline, and it will go to the disk. Either by referencing the could-init data in the disk source section, or by putting the whole cloud-init section there ( (up to david). Both should be good enough, right?

michalskrivanek replied:
sure, both is fine. keeping it separate(referencing) might be better for other non-disk ways of specifying cloudinit, like openstack metadata service. But it doesn't really matter much indeed

fabiand commented on pkg/api/v1/schema.go:
```diff
@@ -25,6 +25,20 @@ package v1
 */
 
+// http://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html
+type CloudInitDataSourceNoCloud struct {
```
Documentation for all fields is needed.

fabiand commented on pkg/api/v1/schema.go:
```diff
@@ -25,6 +25,20 @@ package v1
+}
+
+type CloudInitSpec struct {
+	DataSource string `json:"dataSource"`
```
Documentation for this field is needed.

davidvossel replied:
:+1:

fabiand commented on pkg/api/v1/schema.go:
```diff
@@ -25,6 +25,20 @@ package v1
+type CloudInitSpec struct {
+	DataSource string `json:"dataSource"`
+
+	// DataSource specific data structures
```
Pleas improve the documentation

davidvossel replied:
will do :+1:

fabiand commented on pkg/cloud-init/cloud-init.go:
```diff
@@ -0,0 +1,349 @@
+
+	switch vm.Spec.CloudInit.DataSource {
+	case dataSourceNoCloud:
+		filePath := fmt.Sprintf("%s/%s", GetDomainBasePath(domain, namespace), noCloudFile)
```
Security risk as domain is pretty unfiltered user-input. You might want to use the uuid instead.

davidvossel replied:
domain and namespace shouldn't be allowed to be structured in anyway that would present a security risk here.

As a safety precaution we could return an error if any characters are present that are not alphanumeric, '.' or '-'. Those are the only characters that should ever be in the namespace or domain, so under normal operation that safety check should never get triggered.

fabiand replied:
I hesitate to set a restriction on the name of a domain, because we are using it to create some paths.

We have alternatives which don't suffer from this: Some kind of encoding or using the uuid (which is effectively a charset limited name :) )

davidvossel replied:
> I hesitate to set a restriction on the name of a domain, because we are using it to create some paths.

We aren't making up the restrictions to adhere to path creation. These restrictions already exist (domain name has to adhere to pod naming restrictions) and happen to be path safe.

rmohr replied:
here some documentation from k8s regarding to allowed characters: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/

btw. also keep in mind that uuids are spatially and temporally unique. In some cases that could lead to a situation where we loose track of a resource on a VM repost (don't know if that would be an issue here).

michalskrivanek replied:
from the link: "By convention, the names of Kubernetes resources should be ...", that's not really a precise limitation, and in the linked https://git.k8s.io/community/contributors/design-proposals/identifiers.md there is no such limitation mentioned nor enforced.

I would like to avoid situation where we choose name to be the unique identifier internally and externally. Users have a tendency to rename "stuff" and kind of expecting the object remain the same.

rmohr replied:
Actually it is enforced on TPRs. We can take the UUIDs in that case (since we use the same UID for the Domain in libvirt like in the k8s cluster). However, that is not always possible. Also here doing clean-ups after a virt-handler resync becomes more difficult, but solvable.

Just to provide more context, since I am not sure if that is clear for people: 

The **primary keys** are always namespace and name. It is not possible in general to query objects by UID in k8s (it might be possible in the future though). Namespace and name are always also internal identifiers. In some cases it matters if the UUID is different (the actual object is a repost), in others not. 

An example where you would not care about different UUIDs: Secrets are referenced on the Pod by name, and when a container starts, it is pulled in by name. If the container is restarted, it will be pulled in by name again, not by UUID. 

An example where a controller cares: In our case, when we re-post a VM, we need to make sure that we construct an appropriate Pod for it. Thus, we need to make sure that the pod on the node really belongs to the VM instance with a specific UUID. Therefore one of our controllers internally tracks the UUIDs too.

An end-user can normally not reference other objects by UUID. Controllers sometimes do that internally.

davidvossel replied:
> Users have a tendency to rename "stuff" and kind of expecting the object remain the same.
the domain name can not change without resulting in a completely new VM.

fabiand commented on pkg/config-disk/config-disk.go:
```diff
@@ -0,0 +1,169 @@
+		}
+		var response *http.Response
+
+		response, err = c.unixClient.Post("http://virtconfigdisk/create/cloudinit/"+namespace+"/"+domain, "application/json", strings.NewReader(string(body)))
```
Issues can occur if domain contains special chars.

davidvossel replied:
The domain can't contain special characters.

The pod requires lower case alphanumeric characters with the addition of '-' and '.'  The vm domain will never get started if it's name doesn't match those requirements, which means virt-handler will never see it here.

fabiand commented on pkg/config-disk/config-disk.go:
```diff
@@ -0,0 +1,169 @@
+
+	var response *http.Response
+	var err error
+	response, err = c.unixClient.Post("http://virtconfigdisk/delete/all/"+namespace+"/"+domain, "", nil)
```
Same here.

fabiand commented on cmd/virt-config-disk/virt-config-disk.go:
```diff
@@ -0,0 +1,45 @@
+
+func main() {
+	logging.InitializeLogging("virt-config-disk")
+	cloudInitDir := flag.String("cloud-init-dir", "/var/run/libvirt/cloud-init-dir", "Base directory for ephemeral cloud init data")
```
Please provide some more documentation somewhere for these two directories - how they differ and what they contai.

davidvossel replied:
yup, will do

davidvossel commented:
> On the VM API level, let's go with a disk based approach, this gives the user more control over how and where the disk is getting attached - Idon't see any drawback of that approach right now

I don't think using a disk approach is the right path forward for cloud-init. 

- There are cloud-init data sources that do not use a disk
- cloud-init is logically separated from the VM spec. They are two separate concepts.  VM spec describes how to launch the VM, cloud-init describes how to provide config data into the vm.  It just so happens that the 'nocloud' datasource manipulates the domain xml to do this at creation time.
- I'd like to avoid having cloud-init data intermingled into the VM spec. By having this data in an isolated spec we can create packages that just know how to interpret cloud-init data without having to scan through a VM spec for special cloud-init 'hints'.  

Q: Is there a limit on the base64 encoded content?

nope, not right now.

fabiand commented:
> @fabiand I don't think using a disk approach is the right path forward for cloud-init.
> 
>     There are cloud-init data sources that do not use a disk

Right - We could say that this implementation - if on the VM level - is for noCloud only.

>     cloud-init is logically separated from the VM spec. They are two separate concepts. VM spec describes how to launch the VM, cloud-init describes how to provide config data into the vm. It just so happens that the 'nocloud' datasource manipulates the domain xml to do this at creation time.

Not sure, it sounds we miss each other here - I'm also saying that I do not want to mix the domain and cloud init parts. I'm rather suggesting to explicitly separate them even more (by having domspec in vm.spec.domain, and cloud init in vm.spec.guest.cloudInit).

> Q: Is there a limit on the base64 encoded content?
> 
> nope, not right now.

We should probably do this to avoid performance regressions on a cluster scale.

We also want to at least consider how how users could provide their own disk with custom cloud-init data.
Actually - in the disk case this could be easy - we could either name the source inline, or reference a PVC in case that the data is to large to fit into the spec.

davidvossel commented:
> Not sure, it sounds we miss each other here - I'm also saying that I do not want to mix the domain and cloud init parts. I'm rather suggesting to explicitly separate them even more (by having domspec in vm.spec.domain, and cloud init in vm.spec.guest.cloudInit).

@fabiand I'm not sure I follow what you're proposing then. Can you provide an example of a full VM definition that represents  a VM with a cloud-init nocloud datasource entry in the format you're envisioning so we have something concrete to reference?

davidvossel commented:
Update
- Rolled virt-config-disk component into virt-handler
- Each config-disk request is now executed asynchronously in a go routine.
- Go channels are used for communication instead of a http over a unix socket.

@fabiand  @rmohr These changes simplify the architecture quite a bit. They also allow the ISOs to be created asynchronously to the VM start flow.   The config-disk client remains the generic entry point for executing iso creation jobs, but those jobs are executed in a go routine in the the virt-handler component now (rather than in a separate component with communication over HTTP). In the future, the config-disk pkg can be expanded to execute sysprep iso creation jobs (or any other configuration based disk standard). The abstraction looks a bit odd right now since cloud-init is the only standard config-disk is supporting initially.

The last thing I need to address in this patch is how the feature is exposed in the VM definition.

davidvossel commented:
retest this please

davidvossel commented:
@fabiand @rmohr Based on our discussion so far, this latest round of patches should be in a mergable state (pending code review)

In this update I moved the CloudInitSpec object into the Disk structure.  Disk based cloud-init datasources are now defined directly in the disk section.

**Example:** cloud-init datasource "NoCloud" as a disk.  The source file for the file disk is dynamically generated and injected into the disk's Source field during the VM start flow. 
```
metadata:
  name: testvm-nocloud
apiVersion: kubevirt.io/v1alpha1
kind: VM
spec:
  domain:
    devices:
      disks:
      - type: file
        cloudinit:
            nocloud:
                userDataBase64: I2Nsb3VkLWNvbmZpZwpwYXNzd29yZDogYXRvbWljCnNzaF9wd2F1dGg6IFRydWUKY2hwYXNzd2Q6IHsgZXhwaXJlOiBGYWxzZSB9Cg==
        target:
          dev: vdb
      - type: ContainerRegistryDisk:v1alpha
        source:
          name: kubevirt/cirros-registry-disk-demo:devel
        target:
          dev: vda      
      interfaces:
      - source:
          network: default
        type: network
    memory:
      unit: MB
      value: 64
    os:
      type:
        os: hvm
    type: qemu

```

I did not include enforcement of a userdata limit in these patches. I plan to follow up with that work in a separate PR once we get this initial cloud-init support merged. Issue #404 is tracking the userdata limit. 

also, the functional tests pass for me. The CI failure appears to be happening during setup before the tests can even run.

rmohr commented:
retest this please

stu-gott commented on pkg/cloud-init/cloud-init.go:
```diff
@@ -0,0 +1,376 @@
+
+	uid, err := strconv.Atoi(usrObj.Uid)
+	if err != nil {
+		return err
```
It might be helpful throughout this module to also log this type of error before returning it.

davidvossel replied:
The error gets bubbled up through the config-disk package into virt-handler during the VM launch flow.  The error is in the reason field for "Synchronizing the VM failed." log message.

It wouldn't hurt to add a log message here in the package with a lesser log priority though.

stu-gott commented on pkg/cloud-init/cloud-init.go:
```diff
@@ -0,0 +1,376 @@
+		userData64 := spec.NoCloudData.UserDataBase64
+		metaData64 := spec.NoCloudData.MetaDataBase64
+
+		os.Remove(userFile)
```
It's theoretically possible errors besides "no such file or directory" could occur here. e.g. "permission denied". Worth checking?

davidvossel replied:
yeah, I debated about that.

My thought process here is I don't want to to block creating the iso and starting the VM because of being unable to cleanup these temporary files. Even if we can't clean of the temporary files here during the 'Generate' function, they will get cleaned up later on when the entire folder is deleted as a result of the VM shutting down.

so... I could check for errors, but I think I'd just log that they occurred rather than taking any action. I don't feel very strongly one way or the other.

stu-gott replied:
+1 on logging. My concern is that if there's a (completely unexpected) issue with cleanup before starting, that doesn't bode well for the rest of the process, so making this issue visible is worth doing.

stu-gott commented on pkg/cloud-init/cloud-init.go:
```diff
@@ -0,0 +1,376 @@
+		metaFile := fmt.Sprintf("%s/%s", domainBasePath, "meta-data")
+		userFile := fmt.Sprintf("%s/%s", domainBasePath, "user-data")
+		iso := fmt.Sprintf("%s/%s", domainBasePath, noCloudFile)
+		isoStaging := fmt.Sprintf("%s/%s.staging", domainBasePath, noCloudFile)
```
Suggest that the staging file should be a dotfile, but that's largely a matter of preference. The only hard requirement here is that the two files need to be in the same directory--which you already did.

davidvossel replied:
yup, not a bad idea. I'll update the patch with this.

davidvossel replied:
i don't care strongly about this one way or the other. I ended up leaving this logic alone.

stu-gott commented:
As far as the base64 data, it looks like a mix of yaml and json. Just curious, how is that part generated?

davidvossel commented:
> As far as the base64 data, it looks like a mix of yaml and json. Just curious, how is that part generated?

Are you talking about how the base64 userdata is generated? The contents of that userdata can be all kinds of things. Cloud-init has this cloud-config-data standard that can be used which is defined in yaml, http://cloudinit.readthedocs.io/en/latest/topics/format.html#cloud-config-data

The userdata could also be a bash script, http://cloudinit.readthedocs.io/en/latest/topics/format.html#user-data-script

The userdata is something we just pass through to the cloud-init datasource we support. The cloud-init scripts on the VM are what make sense out of the user-data and execute it on startup.  People defining the user-data have to conform to some method of structuring the user-data that cloud-init can consume. KubeVirt doesn't enforce that part though.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -181,13 +196,18 @@ func (d *VMHandlerDispatch) Execute(store cache.Store, queue workqueue.RateLimit
+	} else if isPending {
+		// waiting on an async action to complete
+		logging.DefaultLogger().V(3).Info().Object(vm).Reason(err).Msg("Synchronizing is in a pending state.")
+		queue.AddRateLimited(key)
```
I think you should use something like `queue.AddAfter(key, 1 * time.Second)`, followed by `queue.Forget(key)`, to cleare the rate limiter.

If we reach this point we had a successful controller run, we are just not yet done.

davidvossel replied:
yup, good idea. i fixed this.

rmohr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -103,9 +111,11 @@ func main() {
 		panic(err)
 	}
 
+	configDiskClient := configdisk.NewConfigDiskClient()
```
On controller-startup, after all chaches are synchronized, we have to clean the disks for all VMs which are in a final state or which have disappeared form the cluster. It should be possible to create a "Reference" to a VM (a VM object only with name and namespace set), based on the directories. Then `disks.Undefine(vmref)` can be invoked for every one of it. 

We would clean the volumes for all VMs which:

 * Are in a final state or removed from the cluster, and
 * Are not currently running on the node (since we are in the synchronization phase, we want to avoid deleting volumes of not yet stopped domains)

davidvossel replied:
Another good catch!

It should be enough to remove directories for any VM that isn't in the cluster still. VMs still known by libvirt in any state will get cleaned up eventually.

davidvossel commented on pkg/config-disk/config-disk.go:
```diff
@@ -0,0 +1,102 @@
+	defer c.lock.Unlock()
+	namespace := precond.MustNotBeEmpty(vm.GetObjectMeta().GetNamespace())
+	domain := precond.MustNotBeEmpty(vm.GetObjectMeta().GetName())
+	delete(c.jobs, namespace+domain)
```
ah, good catch!

davidvossel commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -146,6 +146,11 @@ func main() {
 	vmController.StartInformer(stop)
 	vmController.WaitForSync(stop)
 
+	err = configDiskClient.UndefineUnseen(vmStore)
```
@rmohr  This should handle the startup condition you mentioned where stale data wasn't being cleaned up.

davidvossel commented:
retest this please

davidvossel commented:
Updated, all feedback has been addressed.

davidvossel commented:
retest this please

davidvossel commented:
retest this please

rmohr commented on images/libvirt-kubevirt/Dockerfile:
```diff
@@ -44,4 +44,9 @@ RUN rm -f /libvirtd.sh
+# Make a predictable uid and gid for qemu user
+# This allows containers to set enforce permissions on files that
+# are shared between containers on the local disk.
+RUN groupmod -g 1111 qemu && usermod  -u 1111 qemu
```
We might trip over that later, when we run on OpenShift. In this case we would required the manifest to run with that user.

rmohr commented on pkg/config-disk/config-disk.go:
```diff
@@ -0,0 +1,139 @@
+		if err != nil {
+			return err
+		}
+		_, exists, _ := indexer.GetByKey(key)
```
I think directly using indexer.Get(vm) should also work in this case.

rmohr commented on tests/vm_userdata_test.go:
```diff
@@ -0,0 +1,78 @@
+	})
+
+	LaunchVM := func(vm *v1.VM) runtime.Object {
+		obj, err := virtClient.RestClient().Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Get()
```
virtClient.VM(tests.NamespaceTestDefault).Create(vm), can also be used.

rmohr commented on tests/vm_userdata_test.go:
```diff
@@ -0,0 +1,78 @@
+			}
+
+			for idx, vm := range vms {
+				VerifyUserDataVM(vm, objs[idx])
```
What I am missing is a test, which ensures that the no-cloud data is really picked up inside the VM.

I could think of different ways to test this:

 * Create a user via the user-data and log in via a console (There is already a console_test which reads the console)
 * Check if  the prompt appears without the "checking http://169.254.169.254/2009-04-04/instance-id" messages, where it tries to connect to the cloud init server.

Can be done in a follow up patch.

davidvossel replied:
that's a great idea. I'll follow up with that, thanks!
